CORE-6783

MrMatthewLayton opened the pull request:
Added Transaction Component Interfaces and Tests
**NOTE - PR AND MERGE 6781 BEFORE THIS ONE**

MrMatthewLayton committed:
Added Transaction Component Interfaces and Tests

yarivus commented on ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/TransactionComponents.kt:
```diff
@@ -0,0 +1,196 @@
+ * @property REFERENCES_METADATA The references metadata component group. Ordinal = 14.
+ */
+enum class ComponentGroup {
+    TRANSACTION_PARAMETERS, MEMBERSHIP_PARAMETERS, PACKAGES, SIGNATORIES, ATTACHMENTS, NOTARY, TIME_WINDOW, COMMANDS, COMMANDS_METADATA, INPUTS, INPUTS_METADATA, OUTPUTS, OUTPUTS_METADATA, REFERENCES, REFERENCES_METADATA
```
Please break the lines to have each value in a separate line ... more readable ;)

mnesbit commented on ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/TransactionComponents.kt:
```diff
@@ -0,0 +1,196 @@
+@CordaSerializable
+interface TimeWindow {
+    val from: Instant?
+    val until: Instant?
```
Not optional in C5 notaries

mnesbit commented on ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/TransactionComponents.kt:
```diff
@@ -0,0 +1,196 @@
+ * @property REFERENCES The references component group. Ordinal = 13.
+ * @property REFERENCES_METADATA The references metadata component group. Ordinal = 14.
+ */
+enum class ComponentGroup {
```
I don't think this needs to be public (or rather this sort of stuff will end up in data/.. part of corda-api repo as a documentation process not an API you manipulate in flows.

mnesbit commented on ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/TransactionComponents.kt:
```diff
@@ -0,0 +1,196 @@
+
+    val id: SecureHash
+    val size: Int
+    val signatories: Set<PublicKey>
```
Totally forgot about that corner of the API! Will have to remember it when calculating required signers as my mind just assumes set union of keys on commands is sufficient!

mnesbit commented on ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/TransactionComponents.kt:
```diff
@@ -0,0 +1,196 @@
+/**
+ * Defines information that applies to a transaction state.
+ *
+ * @property contractId The class name of the [Contract] associated with the transaction state.
```
class names are ambiguous in OSGi so this is not accurate. It will be an opaque tag that refers to the CPK main bundle plus class name.

mnesbit commented on ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/TransactionComponents.kt:
```diff
@@ -0,0 +1,196 @@
+ */
+@DoNotImplement
+@CordaSerializable
+interface TransactionState<out T : ContractState> {
```
Similar problem to the command signers in that this will need to be split up under the hood to allow partial evidence without sending/deserializing the states. Maybe we allow a synthesized view, but maybe we do just slit this stuff up in some way on the UtxoLedgerTransaction?

mnesbit commented on ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/TransactionComponents.kt:
```diff
@@ -0,0 +1,196 @@
+ */
+@DoNotImplement
+@CordaSerializable
+interface InputOutputGroup<out T : ContractState, out K : Any> {
```
Does anyone use this?

MrMatthewLayton replied:
@mnesbit Not that I know of. Delete?
What I think would be more valuable for `UtxoLedgerTransaction` and `UtxoTransactionBuilder` is a data structure and associated functions to add or get grouped encumbered states, but again, maybe something that isn't required for beta.

mnesbit commented on ledger/ledger-common/src/main/kotlin/net/corda/v5/ledger/common/transaction/Party.kt:
```diff
@@ -0,0 +1,17 @@
+ * @property owningKey The [PublicKey] that represents the current identity.
+ */
+@CordaSerializable
+interface Party {
```
Where do we use this?

yarivus commented on ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/TransactionComponents.kt:
```diff
@@ -0,0 +1,196 @@
+@CordaSerializable
+interface StateAndRef<out T : ContractState> {
+    val state: TransactionState<T>
+    val ref: StateRef
```
I remember that we had a comment about this in the wiki and were going for a different name ...

MrMatthewLayton replied:
We haven't decided on naming of this yet. I've no idea what you'd call it. In Bitcoin it's called an "out point", which is even worse.

yarivus reviewed (commented):
Left some comments ...
I could see that the various tests are very basic as well.

MrMatthewLayton committed:
Merge branch 'release/os/5.0' into MrMatthewLayton/CORE-6783

MrMatthewLayton committed:
Merge branch 'MrMatthewLayton/CORE-6781' into MrMatthewLayton/CORE-6783

# Conflicts:
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/AbstractMockTestHarness.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/BearableStateJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/CommandAndSignatoriesJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/CommandJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/FungibleStateJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/IdentifiableStateJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/IssuableStateJavaApiTests.java

yarivus reviewed (commented):
LGTM.
I saw mostly comments formatting and other formatting changes along with removal of java_package.info files.

omfR3 committed:
CORE-5088: Add Avro schema for consensual ledger persistence requests (#512)

CORE-5088: Add Avro schema for consensual ledger persistence requests

* Added "ConsensualLedgerRequest" Avro schema for consensual ledger persistence requests that supports different loads which determine the type of the persistence action

* Added "FindTransaction" Avro schema for retrieving the specified consensual signed transaction, specified by id

* Added "PersistTransaction" Avro schema for persisting the specified consensual signed transaction

* Added Virtual node schema LEDGER_PERSISTENCE_TOPIC ("virtual.node.ledger.persistence")

MrMatthewLayton committed:
Intermediate commit - switching branches.

benmcmahon100 committed:
CORE-5375: vnode db rollback, new audit table (#608)

Add audit table to keep track of cpi changelogs 

https://github.com/corda/corda-runtime-os/pull/2172

alexey-kadyrov-r3 committed:
CORE-6364: Refactor Merkle tree API (#618)

Refactor the Merkle tree APIs to provide a more consistent interface:

* Separated MerkleTreeHashDigestFactory into two interfaces - one marker for use in application (MerkleTreeHashDigestFactory) and another for internal use (MerkleTreeHashDigestProvider) that needs to be implemented as an extension. Moved the latter to crypto-extensions.
* Moved MerkleTreeFactory to application and created MerkleTreeProvider in the cipher-suite. A separate PR in corda-runtime-os changes MerkleTreeFactoryImpl to proxy into the new MerkleTreeProviderImpl.
* Moved hash digest constants to their own class, HashDigestConstants

Co-authored-by: Oisín Mac Fhearaí <oisin.macfhearai@r3.com>

MrMatthewLayton committed:
All unit tests working correctly.

MrMatthewLayton committed:
Merge branch 'release/os/5.0' into MrMatthewLayton/CORE-6783

yarivus commented on crypto/src/test/java/net/corda/v5/crypto/merkle/HashDigestConstantsJavaApiTest.java:
```diff
@@ -0,0 +1,19 @@
+public class HashDigestConstantsJavaApiTest {
+    @Test
+    public void constantTests() {
+        assertNotNull(HashDigestConstants.HASH_DIGEST_PROVIDER_DEFAULT_NAME);
+        assertNotNull(HashDigestConstants.HASH_DIGEST_PROVIDER_NONCE_NAME);
+        assertNotNull(HashDigestConstants.HASH_DIGEST_PROVIDER_NONCE_VERIFY_NAME);
+        assertNotNull(HashDigestConstants.HASH_DIGEST_PROVIDER_NONCE_SIZE_ONLY_VERIFY_NAME);
+        assertNotNull(HashDigestConstants.HASH_DIGEST_PROVIDER_TWEAKABLE_NAME);
+        assertNotNull(HashDigestConstants.HASH_DIGEST_PROVIDER_LEAF_PREFIX_OPTION);
+        assertNotNull(HashDigestConstants.HASH_DIGEST_PROVIDER_NODE_PREFIX_OPTION);
+        assertNotNull(HashDigestConstants.HASH_DIGEST_PROVIDER_ENTROPY_OPTION);
```
I don't understand the value of this test ..
If those constants are defined and hard-coded why do we need to assert those are not null ?!

yarivus commented on ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/Attachment.kt:
```diff
@@ -0,0 +1,65 @@
+     *
+     * @return Returns the [InputStream] for the current [Attachment].
+     */
+    fun open(): InputStream
```
I think this limits the interface ...
An Attachment can also live in memory or somewhere else than a file (think of building tests for attachments and mocking files for this ... it's harder).

yarivus commented on ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/StateAndRef.kt:
```diff
@@ -0,0 +1,17 @@
+ */
+@DoNotImplement
+@CordaSerializable
+interface StateAndRef<out T : ContractState> {
```
I really think you should create the decided new name for StateAndRef ;)

yarivus commented on ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/TimeWindow.kt:
```diff
@@ -0,0 +1,35 @@
+ */
+@DoNotImplement
+@CordaSerializable
+interface TimeWindow {
```
I'm not following the need for this interface.
Why can't we use a simple Duration object (getting the midpoint and the others) is simple arithmetic in the using/calling code

yarivus reviewed (commented):
@MrMatthewLayton - please see my comments

MrMatthewLayton committed:
Merge branch 'MrMatthewLayton/CORE-6781' into MrMatthewLayton/CORE-6783
